include/nuttx/fs/fs.h

```diff
@@ -206,16 +207,21 @@ struct file_operations
    * treated like unions.
    */
 
-  int     (*close)(FAR struct file *filep);
-  ssize_t (*read)(FAR struct file *filep, FAR char *buffer, size_t buflen);
-  ssize_t (*write)(FAR struct file *filep, FAR const char *buffer,
-                   size_t buflen);
-  off_t   (*seek)(FAR struct file *filep, off_t offset, int whence);
-  int     (*ioctl)(FAR struct file *filep, int cmd, unsigned long arg);
+  int      (*close)(FAR struct file *filep);
+  ssize_t  (*read)(FAR struct file *filep, FAR char *buffer, size_t buflen);
+  ssize_t  (*write)(FAR struct file *filep, FAR const char *buffer,
+                    size_t buflen);
+  off_t    (*seek)(FAR struct file *filep, off_t offset, int whence);
+  int      (*ioctl)(FAR struct file *filep, int cmd, unsigned long arg);
+  int      (*truncate)(FAR struct file *filep, off_t length);
+  FAR void *(*mmap)(FAR struct file *filep, off_t start, size_t length);
+  int      (*munmap)(FAR struct task_group_s *group, FAR struct inode *inode,
```

Comment:
why mummap has group parameter but mmap hasn't? I think mmap need return a context cookie and receive that cookie in munmap, otherwise it's hard to do the cleanup in munmap. Actually, Linux return another ops:
https://github.com/torvalds/linux/blob/master/include/linux/mm.h#L540-L612
We need the similar mechanism to overcome fd may close before unmap.

Reply:
We've been using the mmap returned address and length as the identifier and stored those, together with a reference to the inode in the group-specific list of mappings. If there is really an inode-backed mapping, the inode is the refetrence. It doesn't really then matter whether the fd is open or close when it is being unmapped, the inode can be preserved with the reference counter it already has.

In all normal cases there is no need to pass the "group" as an argument for mmap or munmap; it is always the current group. The only reason why group is passed in munmap is that during process exit, the current group is something else. Doing actual MMU unmapping cannot be done in that case, but there may still be kernel-allocated resources (such as the inode), which need to be freed. So what we have done is that we pass "NULL" as group at task deletion and the current group when we enter munmap normally from "munmap" user call.

Then the driver can do MMU mapping deletion for example by checking "if (group) { }, or such.

We could also pass just some flag, e.g. "bool on_exit", or such, but were thinking that passing the group is more generic.

If you have some clear vision on different interface, please give a suggestion, these are easy to change.

Reply:
It is possible to add the group parameter of course also to the mmap, if you can think of some potential use for it. I would think that trying to mmap in anything else than currently running context would be illegal.

Also, you really need to store the mappings in the group anyways, since the user side only knows about the virtual address and length of the mapping. Also, when you store the mappings in the group, you don't need any "context" specific token; the mmap and unmap are always done in the same currently running context, except in the case of process deletion.

Reply:
> We've been using the mmap returned address and length as the identifier and stored those, together with a reference to the inode in the group-specific list of mappings. If there is really an inode-backed mapping, the inode is the reference. It doesn't really then matter whether the fd is open or close when it is being unmapped, the inode can be preserved with the reference counter it already has.
> 

It's fine for some pseudo file(e.g., char driver) sine there is one to one mapping between inode and hardware instance, but there are other special cases:

1. For real file system, only the mount point bind to an inode, all files inside mount point doesn't have inode at all
2. Some special char drivers(e.g., sock, eventfd, timerfd, signalfd) share one inode with all instances
3. mtd_dev_s doesn't bind to an inode at all

So, the combination is complex.

> In all normal cases there is no need to pass the "group" as an argument for mmap or munmap; it is always the current group. The only reason why group is passed in munmap is that during process exit, the current group is something else. Doing actual MMU unmapping cannot be done in that case (an doesn't need to be done, since the whole address environment is destroyed anyways), but there may still be kernel-allocated resources (such as the inode), which need to be freed. So what we have done is that we pass "NULL" as group at task deletion and the current group when we enter munmap normally from "munmap" user call.
> 
> Then the driver can do MMU mapping deletion for example by checking "if (group) { }, or such.
> 
> We could also pass just some flag, e.g. "bool on_exit", or such, but were thinking that passing the group is more generic.
> 
> If you have some clear vision on different interface, please give a suggestion, these are easy to change.

I think we can let's mmap return a void * as cookie (the implementation decides what should contain in it) and save this cookie together with address and len in task_group, and pass back in munmap.

Or like Linux: https://github.com/torvalds/linux/blob/master/include/linux/fs.h#L2102
pass a new struct(e.g. file_map_s?) which contains all field required to manage the memory map:

1. link list (or rb tree)  node
2. start and length
2. task_group
4. FAR void *priv

With the last field, the implementation could save the private info into priv and retrieve back in munmap, just like how file_s used in open/close callback.

> It is possible to add the group parameter of course also to the mmap, if you can think of some potential use for it. I would think that trying to mmap in anything else than currently running context would be illegal.
> 
> Also, you really need to store the mappings in the group anyways, since the user side only knows about the virtual address and length of the mapping. Also, when you store the mappings in the group, you don't need any "context" specific token; the mmap and unmap are always done in the same currently running context, except in the case of process deletion.

The driver may need store the additional info beside address and length. For example, shm driver may need record the physical page.

Reply:
Yes, we use an implementation of a linked list, stored into the group, which we call "vm_map". I offered an earlier version of this some months ago. When the interfaces for these settle, I can come back to that and offer the "full" unmap solution in another PR.

IMHO, we shouldn't try to copy linux too much. There everything can be buffered/file system cached, and the implementation relies on having full virtual memory management. Although it is great for general purpose OS, I don't think that it is feasible at this point to try to do all of that for NuttX. Instead, let's concentrate for basic needs and keep NuttX as a real-time os :)

Reply:
block_operations and mtd_dev_s need add the similar callback too. BTW, BCH driver need forward the call from file_operations->block_oerations and FTL need forward block_oerations to mtd_s

Reply:
> Yes, we use an implementation of a linked list, stored into the group, which we call "vm_map". I offered an earlier version of this some months ago. When the interfaces for these settle, I can come back to that and offer the "full" unmap solution in another PR.
> 
> IMHO, we shouldn't try to copy linux too much. There everything can be buffered/file system cached, and the implementation relies on having full virtual memory management. Although it is great for general purpose OS, I don't think that it is feasible at this point to try to do all of that for NuttX. Instead, let's concentrate for basic needs and keep NuttX as a real-time os :)

Yes, I agree that we can learn the core design principle from Linux, but the detail and implementation may difference dramatically.

Reply:
I will check this as well.

Maybe I should pull in the "token" struct from the "vm_map", make mmap return a pointer to that ans also give it as an argument to munmap. Then it handles the cases you mentioned above. I will update this PR soonish, latest on monday

Reply:
it's enough to add `FAR void *priv` to `struct vm_map` and pass `vm_map` to mmap/munmap, so the driver can assign priv to the internal allocated struct. BTW, 'file_map_s' look a little better than 'vm_map'.

Reply:
Sure! I still would like to use a name referring generally to virtual memory, and not files, since you will also want to do mappings which are not related to files... (all uses of mmap with NULL file pointer)

Reply:
How about mm_map_s? since not all soc support mmu, but it can still implement/simulate some mmap semantic. I am afraid that vm_map give the reader impression that he can only implement mmap in kernel mode.

Reply:
Sounds good to me!

Reply:
Done; 

I think that with mm_map_s having a "priv" pointer it is better to also create the mm_map_s where the actual mapping is done, and add it to the group's data structure there. Hence it is not needed as a parameter to mmap. So, I added it as a parameter to munmap and mmap just updates the caller's data pointer and returns the error value. This also fits better to current _file_mmap function.

I can provide the actual "munmap" support with mm_map implemented as a linked list soon after we get this PR finished

Reply:
The current design of mmap callback can't return a private pointer, which make the driver is hard to do the dynamic mapping. So, it may better to:
```
struct mm_map_s
{
 struct list_node node;
  FAR void *priv;
  FAR void *start;
  off_t offset;
  size_t length;
  int prot; /* the last two fields are optional */
  int flags;
};

FAR void *mmap(FAR void *start, size_t length, int prot, int flags,
               int fd, off_t offset)
{
  ....
  map->start = start;
  map->offset = offset;
  map->length = length;
  map->prot = prot;
  map->flags = flags;

  ret = file->mmap(&map);
  return map->start;
}
```
 Caller may set prefer address in start field, but callback make the decision and return the final result. The same map can pass to unmap.

Reply:
We can do it like you propose above if you wish; I will just pick the existing mm_map_s structure for starters, since that has been used in a verified solution.

Reply:
mmap/munmap will be implemented by many drivers, it's important to make their prototype in the final/flexible shape, otherwise, we have to modify many drivers again if mmap/munmap change.

Reply:
Yes, agree. It is more flexible to pass the struct. I still left also the file pointer in the i_ops mmap call. It is something that might be needed in the mmap phase by the driver, but not supposed to be stored. Otherwise I think that my existing mm_map_entry_s matches quite well with your proposal, so I kept my own.

Reply:
Yes, you are right, file_s should keep as the first argument.